doc: openthread: Add page showing OT samples memory requirements. (#3095)
nrfconnect/sdk-nrf: doc/nrf/ug_thread_ot_memory.rst

These values are useful to see whether your application has enough space for running on paricular platforms in different OpenThread configurations.

Flash and RAM requirements for nRF52840 and nRF52833
****************************************************

The following tables list flash memory and RAM requirements for the provided samples built using the OpenThread libraries.
The values change depending on the application and hardware platform.

The values in the tables are provided for the CLI example, which works with all OpenThread calls.
For this reason, these values are the best approximation of the memory requirements.

nRF52840 RAM and Flash memory requirements
==========================================

The following table presents memory requrements for samples running on the nRF52840 SoC with the hardware cryptography support provided by the CC310.
+------------+-------------+-----------------+-----------------------------------------+-------------------+--------------------------+---------------------------------------+----------------------+------------------+------------------+

> | Role       | Sample      | Multiprotocol   | Flash OT Stack + BLE Stack + App [KB]   | Flash BLE Stack   | Peristent storage [KB]   | RAM OT Stack + BLE Stack + App [KB]   | RAM BLE Stack [KB]   | Total ROM [KB]   | Total RAM [KB]   |

Author: `Flash BLE Stack` missing unit, will be added in the next commit.

.. _thread_ot_memory:

OpenThread memory requirements
##############################

This page provides information about the layout and the amount of Flash and RAM memory required by the sample using the OpenThread stack.

> These values are useful to see whether your application has enough space for running on paricular platforms in different OpenThread configurations.

Reviewer: ```suggestion
These values are useful to see whether your application has enough space for running on particular platforms in different OpenThread configurations.
```
What platforms and configurations are we talking about here? You mean different boards and feature sets? If so, link to supported boards page and feature set section, please.

Flash and RAM requirements for nRF52840 and nRF52833

> ****************************************************

Reviewer: Currently, looks like no other boards are supported? If so, remove this section and move other headers one level up.

The following tables list flash memory and RAM requirements for the provided samples built using the OpenThread libraries.

> The values change depending on the application and hardware platform.

Reviewer: Is this valid for all samples for Thread in NCS?
By `OpenThread libraries`, you mean the pre-built libraries?

Author: Yes, yes.

The values in the tables are provided for the CLI example, which works with all OpenThread calls.

> For this reason, these values are the best approximation of the memory requirements.

Reviewer: This is a 1:1 copy from nRF5 SDK for TZ. Please update to NCS.
Also, does this mean the CLI sample has the highest memory requirements? Because in line 13 you say these values are valid for all samples.

Author: `Values for the :ref:`Thread CLI sample <ot_cli_sample>`, which works with all OpenThread calls, are only an approximation of the real memory requirements for a CLI application.`

No, this is a real CLI application, so measured requirements **are showing** real memory requirements.

> Samples were compiled with the default :file: `prj.conf`, MTD variant of the CoAP Client has been compiled with the :file: `overlay-mtd.conf`.

Reviewer: ```suggestion
Samples were compiled with the default :file:`prj.conf`, MTD variant of the CoAP Client has been compiled with the :file:`overlay-mtd.conf`.
```
Maybe an introductory section "How the values were compiled" would be good for information in lines 13-32?
You could also explain in such section why the samples were compiles with these variants.

Below options were used to enable the multiprotocol in chosen configurations:

> :option:`CONFIG_MPSL=y`

Reviewer: ```suggestion
* :option:`CONFIG_MPSL` set to `y`
```
This applies to the whole list below. Otherwise, Kconfig reference links are not working.

:option:`CONFIG_BT_LL_SOFTDEVICE_DEFAULT=y`
:option:`CONFIG_BT=y`
:option:`CONFIG_BT_SMP=y`
:option:`CONFIG_BT_PERIPHERAL=y`
:option:`CONFIG_BT_DEVICE_NAME="NUS_CoAP_client"`
:option:`CONFIG_BT_DEVICE_APPEARANCE=833`
:option:`CONFIG_BT_MAX_CONN=1`
:option:`CONFIG_BT_MAX_PAIRED=1`
:option:`CONFIG_BT_GATT_NUS=y`
:option:`CONFIG_BT_SETTINGS=y`

> nRF52840 RAM and Flash memory requirements

Reviewer: Board names should match what is in https://developer.nordicsemi.com/nRF_Connect_SDK/doc/latest/nrf/gs_programming.html#board-names

So either `nRF52840 DK` or `nrf52840dk_nrf52840`.
Same goes for 833 below.
```suggestion
nRF52840 DK RAM and flash requirements
```

==========================================

> The following table presents memory requrements for samples running on the nRF52840 SoC with the hardware cryptography support provided by the CC310.

Reviewer: ```suggestion
The following table lists memory requirements for samples running on the nRF52840 DK hardware platform with the hardware cryptography support provided by the CC310.
```

> +------------+-------------+-----------------+---------------------------+----------------------+--------------------------+---------------------------+----------------------+------------------+------------------+

Reviewer: The table formatting is copied from nRF5 SDK for TZ, and needs to follow NCS style: https://developer.nordicsemi.com/nRF_Connect_SDK/doc/latest/nrf/doc_styleguide.html#tables

.. _thread_ot_memory:

OpenThread memory requirements
##############################

This page provides information about the layout and the amount of flash memory and RAM that is required by sample that are using the OpenThread stack.
Use these values to see whether your application has enough space for running on particular platforms for different application configurations.

.. _thread_ot_memory_introduction:

How to read the tables
**********************

The memory requirement tables list flash memory and RAM requirements for samples that were built using the available :ref:`nrfxlib:ot_libs`.

The values change depending on the sample, device type, and hardware platform.
Moreover, take into account the following considerations:

* All samples were compiled using the default :file:`prj.conf` file and the master image from :ref:`nrfxlib:ot_libs`.
* The MTD variant of the :ref:`Thread CoAP Client <coap_client_sample>` for the End Device role was compiled with the :file:`overlay-mtd.conf` file.
* To enable the multiprotocol support, the following options were used:

>   * :option:`CONFIG_MPSL` set to ``y``

Reviewer 2: Maybe add a note that this is the default value for all samples?

| Role       | Sample                                  | Multiprotocol   |   ROM OT stack + app [kB] |   ROM Bluetooth LE stack [kB] |   Persistent storage [kB] |   RAM OT stack + app [kB] |   RAM Bluetooth LE stack [kB] | Total ROM [kB]    |   Total RAM [kB] |
+============+=========================================+=================+===========================+===============================+===========================+===========================+===============================+===================+==================+
| Router     | :ref:`CLI <ot_cli_sample>`              | No              |                       455 |                             0 |                        32 |                        97 |                             0 | 487               |               97 |
+------------+-----------------------------------------+-----------------+---------------------------+-------------------------------+---------------------------+---------------------------+-------------------------------+-------------------+------------------+

> | Router     | :ref:`CLI <ot_cli_sample>`              | Yes             |                       583 |                           128 |                        32 |                        97 |                            17 | 743 (exceeds ROM) |              114 |

Reviewer 2: What about two new columns showing the available flash and RAM? Then the `(exceeds ROM)` comment can be removed.
